pallets/api/src/messaging/mod.rs

```diff
@@ -403,49 +443,68 @@ pub mod pallet {
 		pub fn xcm_response(
 			origin: OriginFor<T>,
 			query_id: QueryId,
-			response: Response,
+			xcm_response: Response,
 		) -> DispatchResult {
 			T::XcmResponseOrigin::ensure_origin(origin)?;
-
-			// Lookup message from query id.
-			let (origin, id) = XcmQueries::<T>::get(query_id).ok_or(Error::<T>::InvalidQuery)?;
-			let Some(Message::XcmQuery { query_id, callback, deposit, status }) =
-				Messages::<T>::get(&origin, &id)
-			else {
-				return Err(Error::<T>::InvalidMessage.into())
+			let (origin, id) = XcmQueries::<T>::get(query_id).ok_or(Error::<T>::MessageNotFound)?;
+			let xcm_query_message =
+				Messages::<T>::get(&origin, &id).ok_or(Error::<T>::MessageNotFound)?;
+
+			let (query_id, callback, deposit) = match &xcm_query_message {
+				Message::XcmQuery { query_id, callback, deposit } => (query_id, callback, deposit),
+				Message::XcmTimeout { .. } => return Err(Error::<T>::RequestTimedOut.into()),
+				_ => return Err(Error::<T>::InvalidMessage.into()),
 			};
 
-			// Attempt callback with response if specified.
+			// Emit event before possible callback execution.
+			Self::deposit_event(Event::<T>::XcmResponseReceived {
+				dest: origin.clone(),
+				id,
+				query_id: *query_id,
+				response: xcm_response.clone(),
+			});
+
 			if let Some(callback) = callback {
-				log::debug!(target: "pop-api::extension", "xcm callback={:?}, response={:?}", callback, response);
-				if Self::call(origin.clone(), callback, id, &response, deposit).is_ok() {
-					Self::deposit_event(Event::<T>::XcmResponseReceived {
-						dest: origin,
-						id,
-						query_id,
-						response,
-					});
-					return Ok(());
+				// Attempt callback with response if specified.
+				log::debug!(target: "pop-api::extension", "xcm callback={:?}, response={:?}", callback, xcm_response);
+				match Self::call(&origin, callback.to_owned(), &id, &xcm_response) {
+					Ok(_) => {
+						Messages::<T>::remove(&origin, &id);
+						XcmQueries::<T>::remove(query_id);
+						T::Deposit::release(
+							&HoldReason::Messaging.into(),
+							&origin,
+							*deposit,
+							Exact,
+						)?;
+					},
+					Err(_) => {
+						// Due to dependance on runtime implementation this is hard to unit test.
+						// TODO: Discuss with peter options
```

Comment:
I may not fully understand the comment, but couldn't you create a mock implementation `CallbackExecutor` that you can force a failure with?

Reply:
This has been handled in the callback executor PR and is unrelated to this PR, i will remove the comment for now.